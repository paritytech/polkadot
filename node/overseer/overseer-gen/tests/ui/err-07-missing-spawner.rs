#![allow(dead_code)]

use polkadot_overseer_gen::*;

#[derive(Default)]
struct AwesomeSubSys;

impl ::polkadot_overseer_gen::Subsystem<OverseerSubsystemContext<MsgStrukt>, OverseerError> for AwesomeSubSys {
	fn start(self, _ctx: OverseerSubsystemContext<MsgStrukt>) -> SpawnedSubsystem<OverseerError> {
		unimplemented!("starting yay!")
	}
}

#[derive(Clone, Debug)]
pub struct SigSigSig;

pub struct Event;

#[derive(Clone, Debug)]
pub struct MsgStrukt(u8);

#[overlord(signal=SigSigSig, error=OverseerError, event=Event, gen=AllMessages)]
struct Overseer {
	#[subsystem(MsgStrukt)]
	sub0: AwesomeSubSys,
	i_like_pie: f64,
}

#[derive(Debug, Clone)]
pub struct DummySpawner;

impl SpawnNamed for DummySpawner {
	fn spawn_blocking(
		&self,
		task_name: &'static str,
		subsystem_name: Option<&'static str>,
		_future: futures::future::BoxFuture<'static, ()>,
	) {
		unimplemented!("spawn blocking {} {}", task_name, subsystem_name.unwrap_or("default"))
	}

	fn spawn(
		&self,
		task_name: &'static str,
		subsystem_name: Option<&'static str>,
		_future: futures::future::BoxFuture<'static, ()>,
	) {
		unimplemented!("spawn {} {}", task_name, subsystem_name.unwrap_or("default"))
	}
}

struct DummyCtx;

fn main() {
	let _ = Overseer::builder()
		.sub0(AwesomeSubSys::default())
		.i_like_pie(std::f64::consts::PI)
		//.spawner(DummySpawner) // Spawner is missing
		.build()
		.unwrap();
}
