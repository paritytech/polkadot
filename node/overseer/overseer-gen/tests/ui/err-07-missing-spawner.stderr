error[E0599]: no method named `build` found for struct `OverseerBuilder<Missing<_>, Init<AwesomeSubSys>, Init<f64>>` in the current scope
  --> tests/ui/err-07-missing-spawner.rs:59:4
   |
22 | #[overlord(signal=SigSigSig, error=OverseerError, event=Event, gen=AllMessages)]
   | -------------------------------------------------------------------------------- method `build` not found for this
...
59 |         .build()
   |          ^^^^^ method not found in `OverseerBuilder<Missing<_>, Init<AwesomeSubSys>, Init<f64>>`
   |
   = note: the method was found for
           - `OverseerBuilder<Init<S>, Init<AwesomeSubSys>, Init<f64>>`
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `build`, perhaps you need to implement one of them:
           candidate #1: `frame_support::traits::hooks::GenesisBuild`
           candidate #2: `prometheus::vec::MetricVecBuilder`
